core/iwasm/interpreter/wasm_loader.c

```diff
@@ -12083,6 +12083,14 @@ wasm_loader_prepare_bytecode(WASMModule *module, WASMFunction *func,
 
                 read_leb_uint32(p, p_end, type_idx);
 #if WASM_ENABLE_REF_TYPES != 0 || WASM_ENABLE_GC != 0
+#if WASM_ENABLE_WAMR_COMPILER != 0
+                if (*p != 0x00) {
+                    // Any non-0x00 byte requires the ref types proposal.
+                    // This is different from checking the table_idx value
+                    // since `0x80 0x00` etc. are all valid encodings of zero.
+                    module->is_ref_types_used = true;
```

Comment:
when reference-types is enabled, the encoding difference (0x08 0x00 and 0x00) should not make the output different. i'm not sure if it's a good idea to flip the feature bit (WASM_FEATURE_REF_TYPES) depends on the leb encoding.

how about:
* if reference-types is disabled by the wamrc option, make the validation (wasm_loader.c) fail on 0x80 0x00
* if reference-types is enabled, you don't need to set is_ref_types_used here because the encoding difference should not affect the aot compilation.
* given that the module is already validated, another parser (aot_compile_func) can always parse it as leb128.

Reply:
> given that the module is already validated, another parser (aot_compile_func) can always parse it as leb128.

For the last bullet point, do you mean that I should instead update `aot_compiler.c` to always parse this field as leb128 instead of checking `comp_ctx->enable_ref_types`? Afaict that should be sufficient. More concretely, remove the conditional in https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/compilation/aot_compiler.c#L1171-L1177.

> if reference-types is disabled by the wamrc option, make the validation (wasm_loader.c) fail on 0x80 0x00

I believe the logic in `wasm_loader.c` currently only has access to `WASMModule` and `WASMFunction` but not to the options object. Would this require some refactoring to thread that option into `wasm_loader.c`? It looks like wamrc currently calls the loader pretty early on without passing any options (beside filename) but there appears to be `wasm_runtime_load_ex` with `LoadArgs` that could be loaded to forward that signal..?

Reply:
Yes, currently the aot compiler's options cannot be passed and handled in wasm loader, one workaround that I can find out is to make wamrc bail out as early as we can when creating the aot compiler context, like the handling of SIMD options:
```C
diff --git a/core/iwasm/compilation/aot_llvm.c b/core/iwasm/compilation/aot_llvm.c
index fcf29191..7dac3e57 100644
--- a/core/iwasm/compilation/aot_llvm.c
+++ b/core/iwasm/compilation/aot_llvm.c
@@ -3108,6 +3108,16 @@ aot_create_comp_context(const AOTCompData *comp_data, aot_comp_option_t option)
         goto fail;
     }

+    /* Return error if ref-types and GC are disabled by command line but
+       ref-types instructions are used */
+    if (!option->enable_ref_types && !option->enable_gc
+        && wasm_module->is_ref_types_used) {
+        aot_set_last_error("ref-types instruction was found, "
+                           "try removing --disable-ref-types option "
+                           "or adding --enable-gc option.");
+        goto fail;
+    }
+
     /* Disable features when they are not actually used */
     if (!wasm_module->is_simd_used) {
         option->enable_simd = comp_ctx->enable_simd = false;
```

Reply:
> > given that the module is already validated, another parser (aot_compile_func) can always parse it as leb128.
> 
> For the last bullet point, do you mean that I should instead update `aot_compiler.c` to always parse this field as leb128 instead of checking `comp_ctx->enable_ref_types`?

yes.

Reply:
> Yes, currently the aot compiler's options cannot be passed and handled in wasm loader, one workaround that I can find out is to make wamrc bail out as early as we can when creating the aot compiler context, like the handling of SIMD options:
> 
> ```c
> diff --git a/core/iwasm/compilation/aot_llvm.c b/core/iwasm/compilation/aot_llvm.c
> index fcf29191..7dac3e57 100644
> --- a/core/iwasm/compilation/aot_llvm.c
> +++ b/core/iwasm/compilation/aot_llvm.c
> @@ -3108,6 +3108,16 @@ aot_create_comp_context(const AOTCompData *comp_data, aot_comp_option_t option)
>          goto fail;
>      }
> 
> +    /* Return error if ref-types and GC are disabled by command line but
> +       ref-types instructions are used */
> +    if (!option->enable_ref_types && !option->enable_gc
> +        && wasm_module->is_ref_types_used) {
> +        aot_set_last_error("ref-types instruction was found, "
> +                           "try removing --disable-ref-types option "
> +                           "or adding --enable-gc option.");
> +        goto fail;
> +    }
> +
>      /* Disable features when they are not actually used */
>      if (!wasm_module->is_simd_used) {
>          option->enable_simd = comp_ctx->enable_simd = false;
> ```

i guess it works, yes.
for a longer term, it might be simpler to pass the enabled feature bits to the loader though.

Reply:
@yamt I merged this PR and submitted PR #3726 to fix more issues, please help have a look when you are available. Thanks.

Reply:
Thanks for the fast feedback and sorry for my delays in responding to it! :)

Reply:
@jkrems welcome, it doesn't matter😊